src/ruby/puppet-server-lib/puppet/server/master.rb

```diff
@@ -57,6 +60,53 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def getClassInfoForAllEnvironments()
+    # The clear_environment_settings(env) ensures that the calls which happen
+    # later on to enumerate all of the manifests for each environment are
+    # using the latest environment.conf settings for each environment.  Without
+    # this call, cached (and, therefore, possibly stale) environment.conf
+    # settings would be used.  We want to return the latest data for any calls
+    # made to this method.
+    @env_loader.list.each do |env|
+      Puppet.settings.clear_environment_settings(env)
+    end
+
+    environments =
+      Hash[@env_loader.list.collect do |env|
+       [env.name, self.class.getManifests(env)]
+      end]
```

Comment:
Ok, I agree with you, @cprice404, that having an API into core Ruby Puppet that Puppet Server would call into would greatly simplify and make the logic more maintainable here.

So, @hlindberg, what I would envision is that two new helper methods could be created in core Ruby Puppet code that Puppet Server-level code could call into.  Something like what you did for the `ClassInformationService.classes_per_environment` method would work fine.  The methods would be something like:

* environment_manifests() 

This would require no arguments and when called would return back a `Hash` with keys being the string name of each environment and the value being an `Array` of fully-qualified file names of the manifest files that can be referenced by that environment.

* environment_manifests_for_environment(env_name)

The `env_name` would be a string representing the name of the environment for which manifests files should be found.  The return payload from this call would be an `Array` of fully-qualified names of manifest files that can be referenced by that environment.

--

For both methods above, the list of manifest files returned for an environment should include any manifests that the environment could reference directly - via the directory or file that the `manifest` setting in the environment's `environment.conf` file would either implicitly or explicitly be set to.  The list should also include any manifests that the environment could reference via any of the modules its the `modulepath`.

Also, for both methods above, no "cached" data should be used to compute the result.  No cached values of settings sourced from each environment's `environment.conf` file should be used, e.g., for computing module and manifest paths.  Also, none of the manifests that the module objects return should be from previously cached data.  (I understand that the setting values coming directly from `puppet.conf` will be cached, but I think that's an artifact of Puppet Server's current reloading of global configuration that will need to be looked at outside of the effort for the "NC Class Refresh" epic.) Essentially, we want the result to include the full list of manifest files for each environment as currently defined on disk at the time the method call is made.

In terms of the format of the response data, I'm assuming it should be sufficient for our Puppet Server-level code to just turn right back around and hand the data into the `ClassInformationService.classes_per_environment` method to get parsed class info for each of the manifest files.

--

@hlindberg - does the above make sense?  I can file a PUP ticket with this request and assign to your team for scoping if that makes sense as a next step.